---
title: "ASP-vignette"
author: "Minghao Jiang"
date: "`r Sys.Date()`"
output:
  rmarkdown::html_document:
    highlight: pygments
    toc: true
vignette: >
  %\VignetteIndexEntry{ASP-vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
editor_options: 
  markdown: 
    wrap: 80
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  tidy = FALSE,
  cache = FALSE,
  dev = "png"
)
```

# Packages setup

*Packages messages are suppressed when starting up.*

```{r setup}
pkgs <- c("asp", "glue", "org.Hs.eg.db", "pander")
for (pkg in pkgs) suppressPackageStartupMessages(library(pkg, character.only = T))
```

# Sample files acquisition

A *tx2gene* for each transcript are provided (GENCODE version32).

```{r, results = 'asis'}
tx2gene <- get0("tx2gene", envir = asNamespace("asp"))
pander::pandoc.table(tx2gene[1:10, ], justify = "centre")
```

The raw sample *bam* files and *salmon* results are also provided.

```{r}
bams_s1 <- system.file("extdata", glue("star/sample{1:4}/sample{1:4}.SortedByCoord.bam"), package = "asp")
bams_s2 <- system.file("extdata", glue("star/sample{5:8}/sample{5:8}.SortedByCoord.bam"), package = "asp")
salmons_s1 <- system.file("extdata", glue("salmon/sample{1:4}/"), package = "asp")
salmons_s2 <- system.file("extdata", glue("salmon/sample{5:8}/"), package = "asp")
sampleTable <- data.frame(
  samples = glue("sample{1:8}"),
  conditions = rep(c("s1", "s2"), each = 4),
  files_bam = c(bams_s1, bams_s2),
  dirs_salmon = c(salmons_s1, salmons_s2),
  read_lengths = rep(101, 8),
  library_types = rep("paired-end", 8),
  strandedness = rep("no", 8)
)
print(sampleTable)
```

# ASP object creation

The *tx2gene* flag can be overlooked, and the *gtf* will be used to generate a
*tx2gene* dataframe.

```{r}
asp <- create_asp(
  sampletable = sampleTable,
  tx2gene = tx2gene,
  gtf = "~/doc/reference/gtf/gencode.v32.annotation.gtf",
  gff = "~/doc/reference/gtf/gff/gencode.v32.annotation.gff3",
  fa = "~/doc/reference/fa/GRCh38.p13.genome.fa",
  dir_out = "~/projects/test/",
  conda_path = "~/bin/anaconda3/"
)
```

# Commands generation

You can generate a workflow in **ASP** easily.

```{r}
asp <- cmd_asp(
  asp = asp, conda_env = "rmats", tool = "rmats",
  variable_read_length = T, paired_stats = F, cstat = 0.01, allow_clipping = T
)
```

Multiple workflows can be generated together.

```{r quickStart, eval = FALSE}
asp <- cmd_asp(
  asp = asp, conda_env = c(NA, "assw_3.8"), tool = c("cash", "leafcutter"),
  ##### CASH #####
  cash_jar = "~/bin/cash.jar", MergePval = "G", ram = "100g",
  ##### LeafCutter #####
  leafcutter_dir = "~/bin/leafcutter/"
)
```

# Workflows running

You can set *run* to **FALSE** to check workflows first.

```{r}
asp <- run_asp(asp = asp, tool = "rmats", run = F, read_results = F, parallel = F)
```

Multiple workflows can be performed at the same time using multi sessions.

```{r}
asp <- run_asp(asp = asp, tool = c("cash", "leafcutter"), run = F, read_results = F, parallel = T, block = T)
```

# Results reading

To save the time, we provide the final results outputted by the eight workflows.

```{r}
asp <- create_asp(
  sampletable = sampleTable,
  tx2gene = tx2gene,
  gtf = "~/doc/reference/gtf/gencode.v32.annotation.gtf",
  gff = "~/doc/reference/gtf/gff/gencode.v32.annotation.gff3",
  fa = "~/doc/reference/fa/GRCh38.p13.genome.fa",
  dir_out = system.file("extdata", glue("results/"), package = "asp"),
  nproc = 5,
  novel = F,
  write_log = T,
  parallel = T,
  np = 2,
  conda_path = "~/opt/miniconda3/"
)
asp <- run_asp(
  asp, tool = c("rmats", "cash", "leafcutter", "majiq", "bandits", "psichomics", "suppa", "spladder"),
  run = F, read_results = T, parallel = F
)
```

# Significant differential alternative splicing query

The significant events will be saved in **ASP** through *sig_asp*.

```{r}
asp <- sig_asp(
  asp, tool = c("rmats", "cash", "leafcutter", "suppa", "psichomics"),
  p_adj = 0.05, delta_psi = 0.05
)
asp <- sig_asp(
  asp, tool = "majiq", p_value = 0.05, delta_psi = 0.05
)
asp <- sig_asp(
  asp, tool = "spladder", p_adj = 0.05, effect_size_col = "log2FC_event_count", effect_size_threshod = 1
)
asp <- sig_asp(
  asp, tool = "bandits", p_adj = 0.05, effect_size_col = "DTU_measure", effect_size_threshod = 0.1
)
```

# Significant differential alternative splicing genes intersection and visualization

Intersect the significant differential AS genes expediently.

```{r paged.print = TRUE, results = 'asis'}
asp <- intersect_asp(asp)
pander::pandoc.table(asp@intersection$all[1:5, ])
```

A venn plot and a upset plot can be used to visualize the intersection.
We set venn to FALSE this time. Of course, You can set it to TRUE, and a venn diagram will be added to the top right.

```{r fig.height = 7, fig.width = 8}
venn_upset_asp(asp, venn = F)
```

# Significant differential alternative splicing events visualization

The figures of tracks can be used to check the differences in your data.

```{r fig.height = 7, fig.width = 9}
asp <- plot_pre_asp(
  asp, tool = c("rmats", "cash"), top = 1, samples = 2, txdb = NULL, annotation_db = org.Hs.eg.db
)
plot_asp(asp, tool = "rmats", i = 1)
```

--------------------------------------------------------------------------------

# Session Information

```{r}
sessionInfo()
```
